# Week 5 (3/1-3/7)

## Zoom Recordings

* [Tuesday](https://ub.hosted.panopto.com/Panopto/Pages/Viewer.aspx?id=1fcfd9d8-86f9-47e2-bc9a-acdf01514e74)
* [Thursday](https://ub.hosted.panopto.com/Panopto/Pages/Viewer.aspx?id=3f1ae4e7-bc45-42bb-ab50-ace1014bf97f)

## Notebook

- [View online](../_static/weekly_notebooks/week_5_class.html)
- Download the notebook file: [week_5_class.ipynb](../_static/weekly_notebooks/week_5_class.ipynb).

## [Pandas](../Tools/Pandas/pandas_toc.rst)

* GroupBy
* Index operations
* Combining DataFrames

## Project 

* [Baby names](../Projects/baby_names/baby_names.ipynb)

## Resources

### 1. Sample DataFrame merging data

In [32]:
import pandas as pd
import numpy as np

rng = np.random.default_rng(10)

names = ["Ava", "Benjamin", "Charlotte", "Daniel", "Emma", "Fredric", "Gianna"]
courses = ["MTH 141", "MTH 142", "MTH 241", "MTH 306", "MTH 309", "MTH 311"]
rooms = ["NSC 216", "Capen 110", "Park 440"]

# concat rows data
scores1 = rng.integers(0, 100, 12).reshape(4, 3)
scores2 = rng.integers(0, 100, 9).reshape(3, 3)
columns = ["problem_1", "problem_2", "problem_3"]
sec1 = pd.DataFrame(scores1, index=names[:4], columns=columns)
sec2 = pd.DataFrame(scores2, index=names[4:7], columns=columns)

# concat columns data
scores1 = rng.integers(0, 100, 8).reshape(4, 2)
scores2 = rng.integers(0, 100, 9).reshape(3, 3)
part1 = pd.DataFrame(scores1,
                     index=names[:4],
                     columns=["problem_1", "problem_2"])
part2 = pd.DataFrame(scores2,
                     index=names[:3],
                     columns=["problem_3", "problem_4", "problem_5"])

# merging data
office_nums = rng.integers(100, 150, len(names[:-1]))
courses = pd.DataFrame({"course": courses,
                        "instructor": rng.choice(names[1:], len(courses))})
instructors = pd.DataFrame({"name": names[:-1], "office": office_nums}, dtype="object")

## Exercises (Tuesday)

All exercises below use data on restaurant visits. A DataFrame with this data can be created as follows:

In [23]:
import seaborn as sns
tips = sns.load_dataset("tips")
tips.head(5)

total_bill   tip     sex smoker  day    time  size
0       16.99  1.01  Female     No  Sun  Dinner     2
1       10.34  1.66    Male     No  Sun  Dinner     3
2       21.01  3.50    Male     No  Sun  Dinner     3
3       23.68  3.31    Male     No  Sun  Dinner     2
4       24.59  3.61  Female     No  Sun  Dinner     4

### Exercise 1

How many people visited the restaurant each day of the week? 

**Check:** There were 152 visitors on Thursday.

In [24]:
# "nbsphinx": "hidden"

tips.groupby(by="day")["size"].sum()

day
Thur    152
Fri      40
Sat     219
Sun     216
Name: size, dtype: int64

### Exercise 2

What was the average bill paid each day of the week? 

**Check:** <span class="tex2jax_ignore">The average bill on Thursday was $17.48</span>.

In [25]:
# "nbsphinx": "hidden"

tips.groupby(by="day")["total_bill"].mean()

day
Thur    17.682742
Fri     17.151579
Sat     20.441379
Sun     21.410000
Name: total_bill, dtype: float64

### Exercise 3

For each restaurant visit, compute  what percentage of the total bill was the tip, and then compute the average value of this percentage for male and for female customers. 

**Check:** The average tip for males was 15.76%.

In [26]:
# "nbsphinx": "hidden"

tips["tip_perc"] = 100*tips["tip"]/tips["total_bill"]
tips.groupby(by="sex")["tip_perc"].mean()

sex
Male      15.765055
Female    16.649074
Name: tip_perc, dtype: float64

### Exercise 4

Compute a DateFrame with 10 rows selected from the tips data: 5 rows with the the 5 highest total bills for smokers and 5 rows with the 5 highest total bills for non-smokers


**Check:** The 5 highest bills for smokers were <span class="tex2jax_ignore">$50.81, $43.35, $44.30, $43.11 and $40.55.

In [27]:
# "nbsphinx": "hidden"

def large_bills(df):

    df = df.sort_values(by="total_bill", ascending=False)
    return df.head(5)

big_bills = tips.groupby("smoker").apply(large_bills)
big_bills

total_bill    tip     sex smoker   day    time  size   tip_perc
smoker                                                                     
Yes    170       50.81  10.00    Male    Yes   Sat  Dinner     3  19.681165
       182       45.35   3.50    Male    Yes   Sun  Dinner     3   7.717751
       102       44.30   2.50  Female    Yes   Sat  Dinner     3   5.643341
       197       43.11   5.00  Female    Yes  Thur   Lunch     4  11.598237
       184       40.55   3.00    Male    Yes   Sun  Dinner     2   7.398274
No     212       48.33   9.00    Male     No   Sat  Dinner     4  18.621974
       59        48.27   6.73    Male     No   Sat  Dinner     4  13.942407
       156       48.17   5.00    Male     No   Sun  Dinner     6  10.379905
       142       41.19   5.00    Male     No  Thur   Lunch     5  12.138869
       23        39.42   7.58    Male     No   Sat  Dinner     4  19.228818

### Exercise 5

Create a DataFrame with 4 rows selected from the tips data: a row with the highest total bill for a male smoker, a row with the highest total bill for a male non-smoker, and the same for a female smokers and a female non-smokers. 


**Check:** <span class="tex2jax_ignore">The highest bills were $50.81 for male smoker, $48.33 for a male non-smoker, $44.30 for a female smoker and $35.83 for a female non-smoker.</span>

In [31]:
# "nbsphinx": "hidden"

def max_bills(df):

    return df.loc[df["total_bill"].idxmax()]

big_bills = tips.groupby(["sex", "smoker"]).apply(max_bills)
big_bills

total_bill    tip     sex smoker  day    time  size   tip_perc
sex    smoker                                                                
Male   Yes          50.81  10.00    Male    Yes  Sat  Dinner     3  19.681165
       No           48.33   9.00    Male     No  Sat  Dinner     4  18.621974
Female Yes          44.30   2.50  Female    Yes  Sat  Dinner     3   5.643341
       No           35.83   4.67  Female     No  Sat  Dinner     3  13.033771

### Exercise 6

Create a DataFrame with two rows indicating if a customer was a male or a female, and columns indicating which day of the week the customer visited the restaurant. The value in a given row and column should be the amount of the largest tip that a customer of a given sex and visiting on a given day paid. 

**Check:** <span class="tex2jax_ignore">The highest tip paid on Thursday was $6.70 for males and $5.17 for females.</span>

In [22]:
df.groupby(["sex", "day"])["tip"].max().unstack()

day     Thur   Fri   Sat  Sun
sex                          
Male    6.70  4.73  10.0  6.5
Female  5.17  4.30   6.5  5.2

## Exercises (Thursday)

**Note.** All exercises, except for the first one, use data contained in the `names.zip` file. 

In [150]:
# "nbsphinx": "hidden"

# dataframe with baby names

import zipfile
import pandas as pd

zip = zipfile.ZipFile('names.zip')
files = [n for n in  zip.namelist() if n.endswith(".txt")]
years = [int(f.split(".")[0][-4:]) for f in files]

frames = []
for f in files: 
    with zip.open(f) as foo:
        df = pd.read_csv(foo, names=["name", "sex", "count"])
    frames.append(df)

names = (pd.concat(frames,keys=years)
         .reset_index(level=0)
         .rename({"level_0": "year"}, axis=1)
         .reset_index(drop=True)
        )
names

year       name sex  count
0        1880       Mary   F   7065
1        1880       Anna   F   2604
2        1880       Emma   F   2003
3        1880  Elizabeth   F   1939
4        1880     Minnie   F   1746
...       ...        ...  ..    ...
1989396  2019     Zyheem   M      5
1989397  2019      Zykel   M      5
1989398  2019     Zyking   M      5
1989399  2019        Zyn   M      5
1989400  2019      Zyran   M      5

[1989401 rows x 4 columns]

### Exercise 1

Construct a DataFrame with baby names data coming from the `namesbystate.zip` file. The DataFrame columns should have names "state", "sex", "year", "name", and "count". You don't need to save this data to a single csv file; if you do, the file size will be about 130 MB. 

**Check:** The DataFrame should have 6,122,890 rows.

In [151]:
# "nbsphinx": "hidden"

zip = zipfile.ZipFile('namesbystate.zip')
files = [n for n in  zip.namelist() if n.endswith(".TXT")]
states = [f.split(".")[0] for f in files]

frames = []
for f in files: 
    with zip.open(f) as foo:
        df = pd.read_csv(foo, names=["state", "sex", "year", "name", "count"])
    frames.append(df)

bystate = pd.concat(frames).reset_index(drop=True)
bystate

state sex  year      name  count
0          AK   F  1910      Mary     14
1          AK   F  1910     Annie     12
2          AK   F  1910      Anna     10
3          AK   F  1910  Margaret      8
4          AK   F  1910     Helen      7
...       ...  ..   ...       ...    ...
6122885    WY   M  2019      Theo      5
6122886    WY   M  2019   Tristan      5
6122887    WY   M  2019   Vincent      5
6122888    WY   M  2019    Warren      5
6122889    WY   M  2019    Waylon      5

[6122890 rows x 5 columns]

### Exercise 2

Compute a DataFrame that lists the total number of babies recorded each year. 

**Check:** There were 201,484 babies recorded in 1880 and 3,445,321 in 2019.

In [152]:
# "nbsphinx": "hidden"

n_babies = names.groupby(by="year")["count"].sum()
n_babies

year
1880     201484
1881     192696
1882     221533
1883     216944
1884     243461
         ...   
2015    3694784
2016    3661373
2017    3567022
2018    3503806
2019    3445321
Name: count, Length: 140, dtype: int64

### Exercise 3

Compute a DataFrame that lists the number of male babies named "John" for each year. 

**Check:** There were 9,655 such babies recorded in 1880 and 8,779 in 2019.

In [153]:
# "nbsphinx": "hidden"

n_john = (names[(names["name"] == "John") &
               (names["sex"] == "M")]
               .groupby(by="year")["count"].sum())
n_john

year
1880     9655
1881     8769
1882     9557
1883     8894
1884     9388
        ...  
2015    10396
2016    10032
2017     9499
2018     9163
2019     8779
Name: count, Length: 140, dtype: int64

### Exercise 4

Compute a DataFrame that lists how many different names were used each year for males and how many for females.

**Check:** In 1880 there were 942 different names used for females and 1,058 for males. In 2019 these numbers were 17,905 for females and 14,049 for males.

In [155]:
# "nbsphinx": "hidden"

num_names = names.groupby(by=["year", "sex"])["name"].count()
num_names

year  sex
1880  F        942
      M       1058
1881  F        938
      M        997
1882  F       1028
             ...  
2017  M      14229
2018  F      18100
      M      14062
2019  F      17905
      M      14049
Name: name, Length: 280, dtype: int64

### Exercise 5

Compute a DataFrame that for each name shows in which year the name appeared in the records for the first time.

**Check:** Here are the first recorded years for a few names: Aaban 2007, Aabha 2011, Aabid 2003, Aabidah 2018. 

In [156]:
# "nbsphinx": "hidden"

first_use = names.groupby(by=["name"])["year"].min()
first_use

name
Aaban      2007
Aabha      2011
Aabid      2003
Aabidah    2018
Aabir      2016
           ... 
Zyvion     2009
Zyvon      2015
Zyyanna    2010
Zyyon      2014
Zzyzx      2010
Name: year, Length: 99444, dtype: int64

### Exercise 6

Compute a DataFrame that shows what was the most popular name for males and the most popular name for females each year. 

**Check:** The most popular names in 1880  were John and Mary, and in 2019 Liam and Olivia. 

In [154]:
# "nbsphinx": "hidden"

def top_name(grp):
    return grp.sort_values(by="count", ascending=False).iloc[0]

most_popular = names.groupby(by=["year", "sex"]).agg(top_name)
most_popular\


year    name sex  count
year sex                         
1880 F    1880    Mary   F   7065
     M    1880    John   M   9655
1881 F    1881    Mary   F   6919
     M    1881    John   M   8769
1882 F    1882    Mary   F   8148
...        ...     ...  ..    ...
2017 M    2017    Liam   M  18821
2018 F    2018    Emma   F  18757
     M    2018    Liam   M  19915
2019 F    2019  Olivia   F  18451
     M    2019    Liam   M  20502

[280 rows x 4 columns]